File: lib/data/repository/body_energy_timeline_store.dart
import 'package:drift/drift.dart' show Value;

import '../../core/time/local_date.dart';
import '../../domain/insights/body_energy_timeline.dart';
import '../local/open_vitals_database.dart';

/// How long a past day stays eligible for recomputation.
///
/// Watches sync late and Health Connect back-fills, so a recent day can still
/// gain data and is worth recomputing. A day beyond this window cannot: nothing
/// new will arrive for it, recomputing costs ~8 Health Connect reads to
/// reproduce what is already stored, and — for a user who declined the history
/// grant, which Android cannot offer from the in-app dialog — Health Connect
/// serves only ~30 days, so the recompute can come back empty and take the
/// stored day with it.
///
/// Seven days rather than two or three because a Garmin watch is often synced
/// only once a week, and one sync back-fills every day it covers.
const int bodyEnergyChainSettlingDays = 7;

/// Drift-backed storage for the Body Energy chain: the day summaries whose end
/// scores seed each following day, and the 5-minute buckets behind them.
///
/// Replaces the SharedPreferences timeline cache. That store encoded a whole
/// day as one delimited string, so answering "what did yesterday end on" — the
/// question the chain asks constantly — meant decoding 288 points, and a range
/// query for a multi-day view was impossible.
///
/// The [BodyEnergyTimelineDao] speaks generated rows; the repository speaks
/// [BodyEnergyTimeline]. That mapping lives here so the repository stays about
/// the chain rather than about columns.
class BodyEnergyTimelineStore {
  BodyEnergyTimelineStore(this._dao);

  final BodyEnergyTimelineDao _dao;

  /// The stored timeline for [date], or null when nothing is stored or the
  /// stored row was computed under a different [signature].
  Future<BodyEnergyTimeline?> load(LocalDate date, String signature) async {
    final day = await _dao.day(date.epochDay);
    if (day == null || day.signature != signature) return null;
    final buckets = await _dao.bucketsForDay(date.epochDay);
    return _toTimeline(day, buckets);
  }

  /// Persists [timeline], replacing whatever was stored for its date.
  Future<void> save(BodyEnergyTimeline timeline) async {
    // An unsigned timeline cannot be validated on read, so storing it would
    // only produce a row every future read discards (the prefs store had the
    // same guard).
    if (timeline.signature.trim().isEmpty) return;
    final epochDay = timeline.date.epochDay;
    await _dao.upsertDay(
      _toSummaryCompanion(timeline, epochDay),
      timeline.points
          .map((point) => _toBucketCompanion(point, epochDay))
          .toList(),
    );
  }

  /// The chain-relevant facts for `[start, end]`, oldest first — one query, no
  /// bucket decoding. The walk-back reads its whole lookback window with this.
  Future<List<BodyEnergyStoredDay>> storedDaysBetween(
    LocalDate start,
    LocalDate end,
  ) async {
    final rows = await _dao.daysBetween(start.epochDay, end.epochDay);
    return rows
        .map((row) => BodyEnergyStoredDay(
              date: LocalDate.fromEpochDay(row.epochDay),
              signature: row.signature,
              startScore: row.startScore,
              endScore: row.endScore,
              generatedAt: DateTime.fromMillisecondsSinceEpoch(
                row.generatedAtMillis,
                isUtc: true,
              ),
            ))
        .toList();
  }

  /// Whether a stored day still has buckets behind it.
  ///
  /// Signature-independent on purpose: the caller is protecting the *shape* of
  /// a day it may no longer be able to re-read, and a calibration edit is no
  /// reason to discard that.
  Future<bool> hasStoredPoints(LocalDate date) async =>
      await _dao.countBucketsForDay(date.epochDay) > 0;

  /// Forward ripple: every day in `[from, to]` was computed from a seed that no
  /// longer holds, so drop them and let them be recomputed.
  Future<void> invalidateForward(LocalDate from, LocalDate to) =>
      _dao.deleteDays(from.epochDay, to.epochDay);

  /// Drops the whole chain and its cursor — the calibration/algorithm reset.
  Future<void> purgeAll() => _dao.purgeAll();

  /// Drops buckets older than [bodyEnergyBucketRetentionDays], keeping the day
  /// summaries so the chain stays walkable.
  Future<void> applyRetention(LocalDate today) => _dao.purgeBucketsBefore(
        today.minusDays(bodyEnergyBucketRetentionDays).epochDay,
      );

  /// The global signature the stored chain was built under, or null when the
  /// chain has never been synced.
  Future<String?> storedGlobalSignature() async =>
      (await _dao.chainCursor())?.changesToken;

  Future<void> writeGlobalSignature(String signature) =>
      _dao.writeChainCursor(globalSignature: signature);

  Future<DateTime?> lastPassAt() async {
    final millis = (await _dao.chainCursor())?.lastFullSyncMillis;
    return millis == null
        ? null
        : DateTime.fromMillisecondsSinceEpoch(millis, isUtc: true);
  }

  Future<void> writeLastPassAt(DateTime at) =>
      _dao.writeChainCursor(lastPassMillis: at.toUtc().millisecondsSinceEpoch);
}

/// A stored day's chain-relevant facts.
///
/// Deliberately not a [BodyEnergyTimeline]: the walk-back asks for a fortnight
/// of these on a cold screen open and must not pay a bucket read per day to
/// learn an end score.
class BodyEnergyStoredDay {
  const BodyEnergyStoredDay({
    required this.date,
    required this.signature,
    required this.startScore,
    required this.endScore,
    required this.generatedAt,
  });

  final LocalDate date;
  final String signature;
  final int startScore;
  final int endScore;
  final DateTime generatedAt;
}

BodyEnergyTimeline _toTimeline(
  BodyEnergyDay day,
  List<BodyEnergyBucket> buckets,
) =>
    BodyEnergyTimeline(
      date: LocalDate.fromEpochDay(day.epochDay),
      startScore: day.startScore,
      currentScore: day.endScore,
      charged: day.charged,
      drained: day.drained,
      points: buckets.map(_toPoint).toList(),
      confidence: BodyEnergyConfidence.fromStorage(day.confidence) ??
          BodyEnergyConfidence.noData,
      confidenceReason: day.confidenceReason,
      confidenceReasonCode: bodyEnergyReasonCodeForText(day.confidenceReason),
      inputSummary: BodyEnergyInputSummary(
        algorithmVersion: day.algorithmVersion,
        bucketMinutes: day.bucketMinutes,
        heartRateSampleCount: day.heartRateSampleCount,
        hrvSampleCount: day.hrvSampleCount,
        sleepSessionCount: day.sleepSessionCount,
        workoutCount: day.workoutCount,
        respiratorySampleCount: day.respiratorySampleCount,
        hasRestingHeartRate: day.hasRestingHeartRate,
        hasBaselineRestingHeartRate: day.hasBaselineRestingHeartRate,
        hasObservedMaxHeartRate: day.hasObservedMaxHeartRate,
        hasHrvBaseline: day.hasHrvBaseline,
        hasRespiratoryBaseline: day.hasRespiratoryBaseline,
        previousEndScore: day.previousEndScore,
        carryOverFloorApplied: day.carryOverFloorApplied,
        seedSource: BodyEnergySeedSource.fromStorage(day.seedSource) ??
            BodyEnergySeedSource.neutral,
        calibrationMode:
            BodyEnergyCalibrationMode.fromStorage(day.calibrationMode) ??
                BodyEnergyCalibrationMode.automatic,
      ),
      generatedAt: DateTime.fromMillisecondsSinceEpoch(
        day.generatedAtMillis,
        isUtc: true,
      ),
      signature: day.signature,
    );

BodyEnergyTimelinePoint _toPoint(BodyEnergyBucket bucket) =>
    BodyEnergyTimelinePoint(
      time: DateTime.fromMillisecondsSinceEpoch(bucket.timeMillis, isUtc: true),
      score: bucket.score,
      delta: bucket.delta,
      state: BodyEnergyBucketState.fromStorage(bucket.state) ??
          BodyEnergyBucketState.unmeasurable,
      confidence: BodyEnergyConfidence.fromStorage(bucket.confidence) ??
          BodyEnergyConfidence.noData,
      charge: bucket.charge,
      intensityDrain: bucket.intensityDrain,
      activityEnergyDrain: bucket.activityEnergyDrain,
      basalDrain: bucket.basalDrain,
      stressDrain: bucket.stressDrain,
      recoveryDebtDrain: bucket.recoveryDebtDrain,
      primaryInfluence:
          BodyEnergyPrimaryInfluence.fromStorage(bucket.primaryInfluence) ??
              BodyEnergyPrimaryInfluence.steady,
    );

BodyEnergyDaysCompanion _toSummaryCompanion(
  BodyEnergyTimeline timeline,
  int epochDay,
) {
  final summary = timeline.inputSummary;
  return BodyEnergyDaysCompanion.insert(
    epochDay: Value(epochDay),
    signature: timeline.signature,
    startScore: timeline.startScore,
    endScore: timeline.currentScore,
    charged: timeline.charged,
    drained: timeline.drained,
    confidence: timeline.confidence.storageName,
    confidenceReason: timeline.confidenceReason,
    generatedAtMillis:
        (timeline.generatedAt ?? DateTime.now()).toUtc().millisecondsSinceEpoch,
    algorithmVersion: summary.algorithmVersion,
    bucketMinutes: summary.bucketMinutes,
    heartRateSampleCount: summary.heartRateSampleCount,
    hrvSampleCount: summary.hrvSampleCount,
    sleepSessionCount: summary.sleepSessionCount,
    workoutCount: summary.workoutCount,
    respiratorySampleCount: summary.respiratorySampleCount,
    hasRestingHeartRate: summary.hasRestingHeartRate,
    hasBaselineRestingHeartRate: summary.hasBaselineRestingHeartRate,
    hasObservedMaxHeartRate: summary.hasObservedMaxHeartRate,
    hasHrvBaseline: summary.hasHrvBaseline,
    hasRespiratoryBaseline: summary.hasRespiratoryBaseline,
    previousEndScore: Value(summary.previousEndScore),
    carryOverFloorApplied: summary.carryOverFloorApplied,
    seedSource: summary.seedSource.storageName,
    calibrationMode: summary.calibrationMode.storageName,
  );
}

BodyEnergyBucketsCompanion _toBucketCompanion(
  BodyEnergyTimelinePoint point,
  int epochDay,
) =>
    BodyEnergyBucketsCompanion.insert(
      epochDay: epochDay,
      timeMillis: point.time.toUtc().millisecondsSinceEpoch,
      score: point.score,
      delta: point.delta,
      state: point.state.storageName,
      confidence: point.confidence.storageName,
      charge: point.charge,
      intensityDrain: point.intensityDrain,
      activityEnergyDrain: point.activityEnergyDrain,
      basalDrain: point.basalDrain,
      stressDrain: point.stressDrain,
      recoveryDebtDrain: point.recoveryDebtDrain,
      primaryInfluence: point.primaryInfluence.storageName,
    );

File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/BodyEnergyTimelineStore.kt
package tech.mmarca.openvitals.data.repository

import java.time.Instant
import java.time.LocalDate
import javax.inject.Inject
import javax.inject.Singleton
import tech.mmarca.openvitals.data.local.bodyenergy.BodyEnergyBucketEntity
import tech.mmarca.openvitals.data.local.bodyenergy.BodyEnergyBucketRetentionDays
import tech.mmarca.openvitals.data.local.bodyenergy.BodyEnergyChainCursorKey
import tech.mmarca.openvitals.data.local.bodyenergy.BodyEnergyDayEntity
import tech.mmarca.openvitals.data.local.bodyenergy.BodyEnergyTimelineDao
import tech.mmarca.openvitals.domain.insights.BodyEnergyBucketState
import tech.mmarca.openvitals.domain.insights.BodyEnergyCalibrationMode
import tech.mmarca.openvitals.domain.insights.BodyEnergyConfidence
import tech.mmarca.openvitals.domain.insights.BodyEnergyInputSummary
import tech.mmarca.openvitals.domain.insights.BodyEnergyPrimaryInfluence
import tech.mmarca.openvitals.domain.insights.BodyEnergySeedSource
import tech.mmarca.openvitals.domain.insights.BodyEnergyTimeline
import tech.mmarca.openvitals.domain.insights.BodyEnergyTimelinePoint
import tech.mmarca.openvitals.domain.insights.bodyEnergyReasonCodeForText

/**
 * How long a past day stays eligible for recomputation. Watches sync late
 * and Health Connect back-fills. Seven days, because a Garmin watch is
 * often synced once a week.
 */
const val BodyEnergyChainSettlingDays = 7L

/**
 * Room-backed storage for the Body Energy chain: day summaries and the
 * 5-minute buckets behind them. Maps rows to [BodyEnergyTimeline] so the
 * repository stays about the chain.
 */
@Singleton
open class BodyEnergyTimelineStore @Inject constructor(
    private val dao: BodyEnergyTimelineDao,
) {

    /** The stored timeline for [date], or null when absent or stored under another [signature]. */
    open suspend fun load(date: LocalDate, signature: String): BodyEnergyTimeline? {
        val day = dao.day(date.toEpochDay()) ?: return null
        if (day.signature != signature) return null
        return day.toTimeline(dao.bucketsForDay(date.toEpochDay()))
    }

    /** Persists [timeline], replacing whatever was stored for its date. */
    open suspend fun save(timeline: BodyEnergyTimeline) {
        // An unsigned timeline cannot be validated on read.
        if (timeline.signature.isBlank()) return
        val epochDay = timeline.date.toEpochDay()
        dao.upsertDay(
            timeline.toDayEntity(epochDay),
            timeline.points.map { it.toBucketEntity(epochDay) },
        )
    }

    /** The chain facts for `[start, end]`, oldest first. One query, no bucket decoding. */
    open suspend fun storedDaysBetween(start: LocalDate, end: LocalDate): List<BodyEnergyStoredDay> =
        dao.daysBetween(start.toEpochDay(), end.toEpochDay()).map { row ->
            BodyEnergyStoredDay(
                date = LocalDate.ofEpochDay(row.epochDay),
                signature = row.signature,
                startScore = row.startScore,
                endScore = row.endScore,
                generatedAt = Instant.ofEpochMilli(row.generatedAtMillis),
            )
        }

    /** Whether a stored day still has buckets. Signature-independent on purpose. */
    open suspend fun hasStoredPoints(date: LocalDate): Boolean =
        dao.countBucketsForDay(date.toEpochDay()) > 0

    /** Forward ripple: drop `[from, to]`, whose seeds no longer hold. */
    open suspend fun invalidateForward(from: LocalDate, to: LocalDate) {
        dao.deleteDays(from.toEpochDay(), to.toEpochDay())
    }

    /** Drops the whole chain and its cursor — the calibration/algorithm reset. */
    open suspend fun purgeAll() {
        dao.purgeAll()
    }

    /** Drops buckets older than [BodyEnergyBucketRetentionDays], keeping the summaries. */
    open suspend fun applyRetention(today: LocalDate) {
        dao.purgeBucketsBefore(today.minusDays(BodyEnergyBucketRetentionDays).toEpochDay())
    }

    /** The global signature the stored chain was built under, or null. */
    open suspend fun storedGlobalSignature(): String? =
        dao.cursor(BodyEnergyChainCursorKey)?.changesToken

    open suspend fun writeGlobalSignature(signature: String) {
        dao.writeChainCursor(globalSignature = signature, lastPassMillis = null)
    }

    open suspend fun lastPassAt(): Instant? =
        dao.cursor(BodyEnergyChainCursorKey)?.lastFullSyncMillis?.let(Instant::ofEpochMilli)

    open suspend fun writeLastPassAt(at: Instant) {
        dao.writeChainCursor(globalSignature = null, lastPassMillis = at.toEpochMilli())
    }
}

/** A stored day's chain facts. Not a timeline: the walk-back must not pay a bucket read per day. */
data class BodyEnergyStoredDay(
    val date: LocalDate,
    val signature: String,
    val startScore: Int,
    val endScore: Int,
    val generatedAt: Instant,
)

private fun BodyEnergyDayEntity.toTimeline(
    buckets: List<BodyEnergyBucketEntity>,
): BodyEnergyTimeline =
    BodyEnergyTimeline(
        date = LocalDate.ofEpochDay(epochDay),
        startScore = startScore,
        currentScore = endScore,
        charged = charged,
        drained = drained,
        points = buckets.map { it.toPoint() },
        confidence = confidence.toEnumOrNull<BodyEnergyConfidence>() ?: BodyEnergyConfidence.NO_DATA,
        confidenceReason = confidenceReason,
        // Rows from before reason codes carry only the sentence; unrecognised stays LEGACY.
        confidenceReasonCode = bodyEnergyReasonCodeForText(confidenceReason),
        inputSummary = BodyEnergyInputSummary(
            algorithmVersion = algorithmVersion,
            bucketMinutes = bucketMinutes,
            heartRateSampleCount = heartRateSampleCount,
            hrvSampleCount = hrvSampleCount,
            sleepSessionCount = sleepSessionCount,
            workoutCount = workoutCount,
            respiratorySampleCount = respiratorySampleCount,
            hasRestingHeartRate = hasRestingHeartRate,
            hasBaselineRestingHeartRate = hasBaselineRestingHeartRate,
            hasObservedMaxHeartRate = hasObservedMaxHeartRate,
            hasHrvBaseline = hasHrvBaseline,
            hasRespiratoryBaseline = hasRespiratoryBaseline,
            previousEndScore = previousEndScore,
            carryOverFloorApplied = carryOverFloorApplied,
            seedSource = seedSource.toEnumOrNull<BodyEnergySeedSource>() ?: BodyEnergySeedSource.NEUTRAL,
            calibrationMode = calibrationMode.toEnumOrNull<BodyEnergyCalibrationMode>() ?: BodyEnergyCalibrationMode.AUTOMATIC,
        ),
        generatedAt = Instant.ofEpochMilli(generatedAtMillis),
        signature = signature,
    )

private fun BodyEnergyBucketEntity.toPoint(): BodyEnergyTimelinePoint =
    BodyEnergyTimelinePoint(
        time = Instant.ofEpochMilli(timeMillis),
        score = score,
        delta = delta,
        state = state.toEnumOrNull<BodyEnergyBucketState>() ?: BodyEnergyBucketState.UNMEASURABLE,
        confidence = confidence.toEnumOrNull<BodyEnergyConfidence>() ?: BodyEnergyConfidence.NO_DATA,
        charge = charge,
        intensityDrain = intensityDrain,
        activityEnergyDrain = activityEnergyDrain,
        basalDrain = basalDrain,
        stressDrain = stressDrain,
        recoveryDebtDrain = recoveryDebtDrain,
        primaryInfluence = primaryInfluence.toEnumOrNull<BodyEnergyPrimaryInfluence>() ?: BodyEnergyPrimaryInfluence.STEADY,
    )

private fun BodyEnergyTimeline.toDayEntity(epochDay: Long): BodyEnergyDayEntity =
    BodyEnergyDayEntity(
        epochDay = epochDay,
        signature = signature,
        startScore = startScore,
        endScore = currentScore,
        charged = charged,
        drained = drained,
        confidence = confidence.name,
        confidenceReason = confidenceReason,
        generatedAtMillis = generatedAt.toEpochMilli(),
        algorithmVersion = inputSummary.algorithmVersion,
        bucketMinutes = inputSummary.bucketMinutes,
        heartRateSampleCount = inputSummary.heartRateSampleCount,
        hrvSampleCount = inputSummary.hrvSampleCount,
        sleepSessionCount = inputSummary.sleepSessionCount,
        workoutCount = inputSummary.workoutCount,
        respiratorySampleCount = inputSummary.respiratorySampleCount,
        hasRestingHeartRate = inputSummary.hasRestingHeartRate,
        hasBaselineRestingHeartRate = inputSummary.hasBaselineRestingHeartRate,
        hasObservedMaxHeartRate = inputSummary.hasObservedMaxHeartRate,
        hasHrvBaseline = inputSummary.hasHrvBaseline,
        hasRespiratoryBaseline = inputSummary.hasRespiratoryBaseline,
        previousEndScore = inputSummary.previousEndScore,
        carryOverFloorApplied = inputSummary.carryOverFloorApplied,
        seedSource = inputSummary.seedSource.name,
        calibrationMode = inputSummary.calibrationMode.name,
    )

private fun BodyEnergyTimelinePoint.toBucketEntity(epochDay: Long): BodyEnergyBucketEntity =
    BodyEnergyBucketEntity(
        epochDay = epochDay,
        timeMillis = time.toEpochMilli(),
        score = score,
        delta = delta,
        state = state.name,
        confidence = confidence.name,
        charge = charge,
        intensityDrain = intensityDrain,
        activityEnergyDrain = activityEnergyDrain,
        basalDrain = basalDrain,
        stressDrain = stressDrain,
        recoveryDebtDrain = recoveryDebtDrain,
        primaryInfluence = primaryInfluence.name,
    )

private inline fun <reified T : Enum<T>> String.toEnumOrNull(): T? =
    runCatching { enumValueOf<T>(this) }.getOrNull()
